server/server.go

```diff
@@ -765,3 +808,75 @@ func (s *Server) ID() string {
 	defer s.mu.Unlock()
 	return s.info.ID
 }
+
+// Event subjects
+const (
+	clientConnectEventSubj    = "_SYS.CLIENT.CONNECT"
+	clientDisconnectEventSubj = "_SYS.CLIENT.DISCONNECT"
+)
+
+func (s *Server) sendEventNotification(subject string, v interface{}) {
+
+	// Don't send a notification unless there is interest.
+	subs := s.sl.Match([]byte(subject))
+	if len(subs) <= 0 {
+		return
+	}
+
+	payload, err := json.Marshal(v)
+	if err != nil {
+		Errorf("Error marshaling event notification: %v, %s\n", v, err)
+		return
+	}
+
+	s.eventCli.mu.Lock()
+
+	s.eventCli.processPub([]byte(fmt.Sprintf("%s %d", subject, len(payload))))
+	s.eventCli.processMsg([]byte(fmt.Sprintf("%s\r\n", payload)))
+
+	s.eventCli.flushPendingMessages()
+	s.eventCli.last = time.Now()
+
+	s.eventCli.mu.Unlock()
+}
+
+// caller must lock the client
+func (s *Server) createConnectEvent(c *client, reason string) clientConnEvent {
+
+	c.mu.Lock()
+	nc := c.nc
+	name := c.opts.Name
+	cid := c.cid
+	c.mu.Unlock()
+
+	cda := clientConnEvent{Name: "unknown", IP: "unknown", CID: int(cid), Reason: "unknown"}
```

Comment:
Two things:
- Why not returning a pointer (and so here doing `cda := &clientConnEvent{..`? I am not sure, but this may result in copies being passed around.
- When initializing the structure, you could pass directly `Name: name`, etc... or maybe you couldn't because you do `cda := clientConnEvent{...`?

Even better, you may want to marshal in place and simply return a byte array, that you would pass to the sendNotification function?